microcosm_pubsub/decorators/publish_decorators.py

```diff
@@ -0,0 +1,138 @@
+"""
+Fluent decorators for methods.
+
+"""
+
+from marshmallow import ValidationError
+from functools import wraps
+from microcosm_flask.operations import Operation
+
+from microcosm_pubsub.conventions import created
+
+
+def _get_from_kwargs_or_component(component, key, kwargs):
+    if key in kwargs:
+        return kwargs[key]
+    value = getattr(component, key, None)
+    if value is not None:
+        return value
+    raise ValidationError(f"publish requires {key} argument or component.{key} definition")
+
+
+def _create_default_media_type(component, **kwargs):
+    ns = _get_from_kwargs_or_component(component, "ns", kwargs)
+    return created(ns.subject)
+
+
+def _create_default_uri_extractor(
+    component,
+    operation,
+    uri_string_args=None,
+    **kwargs,
+):
+    ns = _get_from_kwargs_or_component(component, "ns", kwargs)
+    if uri_string_args is None:
+        uri_string_args = dict()
+        identifier_key = _get_from_kwargs_or_component(component, "identifier_key", kwargs)
+        uri_string_args[identifier_key] = lambda model: str(model.id)
+
+    return lambda model: ns.url_for(
+        operation,
+        **{key: str(extractor(model)) for key, extractor in uri_string_args.items()}
+    )
+
+
+def publish(
+    producer_key="sns_producer",
+    media_type=None,
+    default_uri=True,
+    uri_string_args=None,
+    message_params=None,
+    operation=None,
+    **kwargs,
+):
+    """
+    A method decorator generator that publish pubsub messages.
+
+    Simple examples:
+
+    def __init__(self, graph):
+        self.ns = Namespace(subject="company")
+        self.identifier_key = company_id
+
+    @publish()
+    def create(**kwargs):
+        return company(id="ID")
+    produce:
+        application/vnd.globality.pubsub._.created.company message
+        with: {'uri': '/api/v1/company/ID'}
+
+    @publish(media_type=changed("Company"))
+    def create(**kwargs):
+        return company(id="ID")
+    produce:
+        application/vnd.globality.pubsub._.changed.company message
+        with: {'uri': '/api/v1/company/ID'}
+
+    @publish(uri_string_args={"company_id": lambda company: company.id, "limit": lambda company: 1})
+        return company(id="ID")
+    produce:
+        application/vnd.globality.pubsub._.created.company message
+        with: {'uri': '/api/v1/company/ID?limit=1'}
```

Comment:
I'd include an example with the Operation (maybe like the test case where you didn't include uri args).